architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+# Data Loader Design For FMS-HF-Tuning
+
+**Deciders(s)**: Sukriti Sharma (sukriti.sharma4@ibm.com), Will Johnson (Will.Johnson@ibm.com) , Abhishek Maurya (maurya.abhishek@ibm.com), Dushyant Behl (dushyantbehl@in.ibm.com), Ashok Pon Kumar (ashokponkumar@in.ibm.com)
+
+**Date (YYYY-MM-DD)**:  2024-03-06
+
+**Obsoletes ADRs**:  NA
+
+**Modified By ADRs**:  NA
+
+**Relevant Issues**: [1]
+
+- [Summary and Objective](#summary-and-objective)
+  - [Motivation](#motivation)
+  - [User Benefit](#user-benefit)
+- [Decision](#decision)
+  - [Alternatives Considered](#alternatives-considered)
+- [Consequences](#consequences)
+- [Detailed Design](#detailed-design)
+
+## Summary and Objective
+
+The reason for motivating dataloader design for fms-hf-tuning is to have a unified interface which supports many type of data formats, streaming and non streaming data, weight based data mixing and many others. Full list we focus on is below - 
+
+1. Data Loading -
+
+    1. Different formats of data → Arrow, Parquet etc.
+
+    1. Different modalities of data -> Images, Audio etc.
+
+    1. Streaming data.
+
+    1. Data Replay
+
+    1. Resume with different number of GPUs
+
+    1. Async data loading
+
+1. Data Preprocessing - 
+
+    1. Custom Attn Masking
+
+    1. Tool Usage
+
+1. Data Mixing - 
+
+    1. Static weights based mixing
+
+### Motivation
+
+### User Benefit
+
+## Decision
+
+### Alternatives Considered
+
+## Consequences
+
+### Advantages
+
+### Impact on performance
+
+## Detailed Design
+
+The input spec which user specifies on how to pass information to such data loader is this
+
+```
+dataloader:
+    type: default(hfdataloader)/stateful(fms-fsdp implementation)
+    streaming: true
+datasets:
+  - name: dataset1
+    sampling:
+      ratio: 0.3
+    splitter_arguments:
+      test_size: 0
+    data_paths:
+      - /data/stackoverflow-kubectl_posts
+      - /data/stackoverflow-kubernetes_posts
+      - /data/stackoverflow-openshift_posts
+    data_handlers:
+      - name: render_template
+        arguments:
+          remove_columns: all
+          batched: false
+          fn_kwargs:
+            jinja_template: "{<jinja-template>}"
+  - name: dataset1
+    sampling:
+      ratio: 0.4
+    data_paths:
+      - /data/stackoverflow-kubectl_posts
+      - /data/stackoverflow-kubernetes_posts
+      - /data/stackoverflow-openshift_posts
+    splitter_arguments:
+      test_size: 1
+    data_handlers:
+      - name: render_template
+        arguments:
+          remove_columns: all
+          batched: false
+          fn_kwargs:
+            jinja_template: "{<jinja-template>}"
+  - name: dataset2
+    sampling:
+      ratio: 0.3
+    data_handlers:
+      - name: apply_tokenizer_chat_template
```

Comment:
@dushyantbehl @ashokponkumar doesnt SFT trainer add chat template on their own? https://huggingface.co/docs/trl/en/sft_trainer#dataset-format-support if dataset is of certain format . Chat is already supported by SFT trainer, since majority of your data handler examples given by you are applying templates, trying to understand use case